### PR TITLE
Allow to find precise output folder for each language.

### PR DESCRIPTION
**Pull request by sdedic**
In a mixed-language project, such as `micronaut-core/http`, each language produces the output typically to a separate folder, although the output is in form of .class files.

Currently Gradle project knows the set of output folders, but does not know what language produces output in which particular one, so backwards mapping in `SourceForBinaryQuery` provides **all source folders**, which is not correct.

In addition, the result of SFBQ query in Gradle/java says that sources should be always preferred. This is fine in theory, but java uses (exclusively) its cached indexed folders instead of the real output folders - so it is able to react on editing changes even before they are compiled. But other languages behave differently.

These two things in effect lead to spurious "undersolved symbol" when Java code references some Groovy or Kotlin class:
- the `build/groovy/main/classes`, or `build/kotlin/main/classes` is on the `classpath/compile` ClassPath; that's right
- since `preferSources` is true, java support attempts of SFBQ to find sources. It gets a result. 
- And to avoid **parsing**, it replaces the source folder with cache folder, under assumption that the sources are indexed into .sig files Java can process.

The last thing is not true for groovy / kotlin, at least now. This could be improved in Java support if it did not entirely replace the output folder with the source, but used the output folder as a cache for the case the SFBQ result folder and its cache does not contain the desired type.

---
**^Add meaningful description above**

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

### CODE REVIEW

**Review by lkishalmi — APPROVED** [review]
Well it took a while to read through. It looks sane. Thank you for this!
The only thing I would add is to increment the cache version number at 

https://github.com/apache/netbeans/blob/5aafb4fd173a53181aa197d1c8a156d01d71c1fa/extide/gradle/src/org/netbeans/modules/gradle/cache/ProjectInfoDiskCache.java#L48

So it would eventually force reloading projects.

File: java/gradle.java/src/org/netbeans/modules/gradle/java/api/GradleJavaSourceSet.java [hunk 1]
```diff
@@ -365,6 +366,33 @@ public boolean outputContains(File f) {
...
+     * null, if the source type has no output directories. Returns UNKNOWN, if the 
+     * output location is not known.
+     * 
+     * @param srcType language type
+     * @return location or {@code null}.
+     * @since 1.19
+     */
+    public File getOutputClassDir(SourceType srcType) {
+        File f = outputs.get(srcType);
+        if (f == null) {
```
Comment (lbownik) [inline java/gradle.java/src/org/netbeans/modules/gradle/java/api/GradleJavaSourceSet.java:388]: this if seems redundant as UNKNOWN.equals(f) evaluates false it f == null
 so maybe

return UNKNOWN.equals(f) ? UNKNOWN : f;

will do?

File: extide/gradle/netbeans-gradle-tooling/src/main/java/org/netbeans/modules/gradle/tooling/NbProjectInfoBuilder.java [hunk 2]
```diff
@@ -307,6 +354,43 @@ private void detectSources(NbProjectInfoModel model) {
...
+                                                if (candidate != null) {
+                                                    candidate = null;
+                                                    break;
+                                                } else {
+                                                    candidate = p;
+                                                }
+                                            }
+                                        }
+                                    }
+                                    if (candidate != null) {
```
Comment (lbownik) [inline extide/gradle/netbeans-gradle-tooling/src/main/java/org/netbeans/modules/gradle/tooling/NbProjectInfoBuilder.java:384]: how about ternary operator here?

Comment (sdedic) [reply extide/gradle/netbeans-gradle-tooling/src/main/java/org/netbeans/modules/gradle/tooling/NbProjectInfoBuilder.java]: Fixed in 01cb17c54a

File: java/gradle.java/src/org/netbeans/modules/gradle/java/api/GradleJavaSourceSet.java (hunk 1 above)
Comment (sdedic) [reply java/gradle.java/src/org/netbeans/modules/gradle/java/api/GradleJavaSourceSet.java]: Fixed in 01cb17c54a

**Comment by sdedic** [comment]
> 
> So it would eventually force reloading projects.

Fixed in 01cb17c54a

**Review by dbalek — APPROVED** [review]
Looks fine.

**Pull request merged**